### ISSUE TITLE
[curl with http/2 protocol]

### ISSUE THREAD

**Issue by morloderex** [oerdnj/deb.sury.org#596]
Hey Oerdnj

I would like to have curl with HTTP/2 support in both the php extension and libcurl library in order to get my servers up to speed. 

Is this something you can look at?

**Comment by oerdnj** [comment]
Can you elaborate a little more on what's needed on curl side? E.g. what are the extra dependencies?

Also do you have idea whether this changes libcurl ABI?

**Comment by peshi** [comment]
https://curl.haxx.se/docs/http2.html
curl supports http2 via libnghttp2.

**Comment by MikeLund** [comment]
Hi, is this issue under consideration? Ran into this same issue. I want to use HTTP/2 with PHP's curl but isn't possible as far as I can see right now (without recompiling your packages). Thanks!

**Comment by oerdnj** [comment]
Yes, it is on the TODO list, but there only a limited number of new things I am able to add on the portfolio, so please be patient.

**Comment by morloderex** [comment]
Any new on this?

**Comment by morloderex** [comment]
ping @oerdnj

**Comment by oerdnj** [comment]
Ok, so it might happen, but without Ubuntu 14.04 LTS support. That's simply too old and it would require too many dependencies backported.

Does anybody know whether php needs to be recompiled to have HTTP/2 support?

If not, simply adding ppa:ondrej/php-qa might do the trick (17.04 is still compiling, but curl for 16.04 should work now).

**Comment by oerdnj** [comment]
Also please be aware it's basically impossible to do without breaking havoc to some reverse dependencies of libcurl as it exports some OpenSSL symbols. For more background read: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=858398

The package needs SONAME bump, so I will need to wait for Debian Bug #858398 to be fixed first, to have a proper curl with OpenSSL 1.1.x support.

**Comment by oerdnj** [comment]
> Does anybody know whether php needs to be recompiled to have HTTP/2 support?

Nothing in `config.m4`, but there are some `#ifdef`s inside the `interface.c`, so the answer is yes.

**Comment by Ayesh** [comment]
I actually did manage to get curl working with http2 support (Ubuntu 16.10, php 7.0+ curl 7.55). Recompiled and replaced. PHP can now make http/2 curl connections (`CURLOPT_HTTP_VERSION = CURL_HTTP_VERSION_2`)

![curl](https://user-images.githubusercontent.com/811553/33780333-678c8572-dc76-11e7-9283-018ac2abc692.png)

I didn't run into errors so far, but I'm pretty sure there can be bugs.

**Comment by morloderex** [comment]
But that's on a non lts. I managedto recompile CURL from source myself and linked that one to the global libraries and got it working by doing that. But sense curl on 16.04 is version 47 or some older version compiling it with openssl became combersume for me. 

That's why i would like @oerdnj to take care of this. Sense Ubuntu itself wont link nghttp themself.

**Comment by oerdnj** [comment]
@morloderex Please try adding the [ppa:ondrej/php-qa](https://launchpad.net/~ondrej/+archive/ubuntu/php-qa/+packages?batch=300) repository.

**Comment by morloderex** [comment]
I'll try it tonight and report back :) Thank you!

**Comment by morloderex** [comment]
I have no tested this new ppa. With ubuntu 16.04 with no luck updating curl. I have packages conflicts whenever i install this ppa.

**Comment by oerdnj** [comment]
@morloderex You need to add `ppa:ondrej/php-qa` on top of `ppa:ondrej/php` (e.g. you have to use both.

**Comment by oerdnj** [comment]
Otherwise provide the logs about the package conflicts.

**Comment by morloderex** [comment]
Here is the log file of trying to upgrade curl on a new digitalocean running ubuntu 16.04.
https://gist.github.com/morloderex/201a4074816104c0aefd0ca5ec6f7ad6

But as you can see libnghttp2-14 is not the correct version. 

@oerdnj

**Comment by morloderex** [comment]
@oerdnj any news on these package conflicts?

**Comment by oerdnj** [comment]
Guess you need ppa:ondrej/apache2 along with those two.
